### PR TITLE
Add quarkus-oidc-client-registration extension

### LINKED ISSUE #38250: Support for OIDC Dynamic Client Registration
### Description

https://openid.net/specs/openid-connect-registration-1_0.html defines a standard mechanism for registering OIDC clients (tenants).

It will be useful for users who are automating the client/tenant registration. 

Note, Keycloak users can already do it with https://quarkus.io/guides/security-keycloak-admin-client using Keycloak specific API, however there should also be a way to do it for any OIDC provider implementing this spec

### Implementation ideas

Light-weight extension, using Vert.x WebClient, initialized with the OIDC base URL from where it can discover the dynamic registration endpoint address.

It will take a bit of time for me to prioritize but it will be a useful feature to have once it is done

### PR DESCRIPTION
**Pull request by sberyozkin**
Fixes #38250.

This PR introduces `experimental` `quarkus-oidc-client-registration` extension to support OIDC Dynamic Client Registration. Client registration API provided by this extension may change after feedback from users, before it settles down.

OIDC Dynamic Client Registration is an important feature allowing to scale up registration and management of OIDC clients in combination with dynamic OIDC tenancy resolution. 

I'm moving it out of the draft state because this PR is no longer trying to bring 2 new extensions (`quarkus-oidc-client-registration`, `quarkus-oidc-client-registration-configuration-source`) and focuses on the core OIDC client registration only with further enhancements to be added in due course.

There are about 3 updates in this PR which impact a large number of files, but only a single one is really about the OIDC client registration. The other two ones just try to make it easier to move ahead with the OIDC client registration work:

* `quarkus.oidc.runtime.AbstractJsonObject` got moved from `quarkus-oidc` down to `quarkus-oidc-common`, to avoid adding the same methods for supporting retrieving JSON for OIDC client registration `ClientMetadata` (and longer term, to support OIDC client enhancement request from Stian for OIDC client to support UserInfo, etc) - and this move causes a few  import updates
* OIDC client registration is about talking to a discoverable OIDC client registration endpoint, and a lot of current `OidcCommonConfig`, which both `quarkus-oidc` `OidcTenantConfig` and `quarkus-oidc-client` `OidcClientConfig` extend, applies to these communications. But a good portion of `OidcCommonConfig` is only relevant to the work of `quarkus-oidc` `OidcTenantConfig` and `quarkus-oidc-client` `OidcClientConfig`. Therefore this PR moves such properties out of `OidcCommonConfig` to `OidcClientCommonConfig` leading to the following hierarchies: `OidcTenantConfig` -> `OidcClientCommonConfig` -> `OidcCommonConfig`, `OidcClientConfig` -> `OidcClientCommonConfig` -> `OidcCommonConfig`, but only `OidcClientRegistrationConfig` -> `OidcCommonConfig`. This leads to a good number of import updates as well. It is not strictly necessary, but then users who create client registrations would see configuration properties not relevant to the client registration process and that would not be great.

Finally, we have the actual `quarkus-oidc-client-registration` extension. I've tried to provide a summary of registration patterns it supports in this PR's doc update, please have a look. But here is a brief summary:

* Each `OidcClientRegistration` represents an OIDC client registration endpoint - it can be created from `OidcClientRegistrationConfig` set either in `application.properties` or dynamically.
* `OidcClientRegistrationConfig` may specify OIDC address only, but also can have metadata configured for a single client, if it does, then `OidcClientRegistration` will return `RegisteredClient` representing an already registered client at start-up
* Each `OidcClientRegistration`, in addition to returning client registered at startup, may accept requests to register one or more clients, using Uni and Multi response types respectively
* `OidcClientRegistration` can be used without having to use `quarkus-oidc`, but only to facilitate the dynamic client registration
* OIDC dynamic `TenantConfigResolver` is suited best for using dynamically registered clients and creating OIDC tenant configurations
* `OidcClientRegistrations` (as opposed to `OidcClientRegistration`) provides access to all `OidcClientRegistration`s created from configured OIDC client registration configs
* `OidcClientRegistrations` can be used to accept a new `OidcClientRegistration` which in turn can be used to register one or more clients
* `RegisteredClient` represents a registered client and supports interacting with this client-specific registration endpoint to read, update the metadata or delete this client
* `ClientMetadata` is a JSON wrapper which can be used to register new clients

OIDC dynamic client registration is a rather complex area and it is hard to capture all of its subtleties perfectly from the start. Therefore I mark it clearly as an experimental extension, in its metadata and the docs, and I propose to let users stress it after the review dealing with some obvious PR issues. It will become an important feature but it may need a few iterations to settle down.

I'd like to have at least 2 reviews and approvals. Thanks

### CODE REVIEW

File: docs/src/main/asciidoc/security-openid-connect-client-registration.adoc
Comment (gastaldi) [inline docs/src/main/asciidoc/security-openid-connect-client-registration.adoc:30]:
```suggestion
== OIDC Client Registration
```

File: extensions/oidc-client-registration/deployment/src/main/java/io/quarkus/oidc/client/registration/deployment/OidcClientRegistrationBuildTimeConfig.java
Comment (gastaldi) [inline extensions/oidc-client-registration/deployment/src/main/java/io/quarkus/oidc/client/registration/deployment/OidcClientRegistrationBuildTimeConfig.java:16]: Use the interface pattern instead (see https://quarkus.io/guides/writing-extensions#configuration)

File: extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/OidcClientRegistrationConfig.java
Comment (gastaldi) [inline extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/OidcClientRegistrationConfig.java:15]: Use the interface pattern here too (https://quarkus.io/guides/writing-extensions#configuration)

File: extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/runtime/DisabledOidcClientRegistrationException.java
Comment (gastaldi) [inline extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/runtime/DisabledOidcClientRegistrationException.java:4]: A Javadoc here would be helpful

File: extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/runtime/OidcClientConfigurationException.java
Comment (gastaldi) [inline extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/runtime/OidcClientConfigurationException.java:4]: Javadoc missing, when is this exception thrown?

File: extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/runtime/OidcClientRegistrationException.java
Comment (gastaldi) [inline extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/runtime/OidcClientRegistrationException.java:4]: Javadoc missing here too

File: extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/runtime/OidcClientRegistrationImpl.java
Comment (gastaldi) [inline extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/runtime/OidcClientRegistrationImpl.java:176]:
```suggestion
            throw new IllegalStateException("OIDC Client Registration is closed");
```

**Review by gastaldi — APPROVED** [review]
Looks good overall, added some nice-to-have suggestions

File: extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/OidcClientRegistrationConfig.java
Comment (sberyozkin) [reply extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/OidcClientRegistrationConfig.java:15]: We'll need to have #39185 first but if this PR happens to go in earlier than this task will be done as part of #39185. I'm not sure what is the right course of action here.

File: extensions/oidc-client-registration/deployment/src/main/java/io/quarkus/oidc/client/registration/deployment/OidcClientRegistrationBuildTimeConfig.java
Comment (sberyozkin) [reply extensions/oidc-client-registration/deployment/src/main/java/io/quarkus/oidc/client/registration/deployment/OidcClientRegistrationBuildTimeConfig.java:16]: Sure, good point, it will be fixed as part of #39185, but I'm not certain if #39185 must be done first

File: extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/runtime/OidcClientConfigurationException.java
Comment (sberyozkin) [reply extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/runtime/OidcClientConfigurationException.java]: It is a copy and paste side-effect :-), this exception class has gone, thanks for spotting it

**Comment by sberyozkin** [comment]
@michalvavrik, what are your thoughts about an order in which `ConfigMapping` to OIDC and OIDc client extensions, and possibly to the OIDC client registration one should be added ?

This PR introduces `OidcTenantConfig` -> `OidcCommonClientConfig` -> `OidcCommonConfig` hierarchy by moving properties not relevant to OIDC client registration to `OidcCommonClientConfig`, with `OidcClientRegistrationConfig` extending `OidcCommonConfig` directly.

I don't mind much waiting for #39185 before finalizing this PR but completing #39815 may take a while and it would be more difficult afterwards to move some of the builder API support for `OidcCommonConfig`.

If that works for you, then, once everyone is happy with this PR, two of us can look at introducing the builder API for `oidc`, `oidc-client` and `oidc-client-registration`, and for the last one we will just replace `OidcClientRegistrationConfig` group with the mapping without even deprecating the group, since it is an experiemental extension

**Comment by michalvavrik** [comment]
@sberyozkin Users are here instantiating config classes and adjusting them, these classes are part of the public API, I agree you cannot use `@ConfigMapping` them without smart builder. If you can create builder for the client registration only, that it's fine, but given how long discussion for KC AuthZ builder took, I am not sure if you want to do this as part of this PR, because new mapper needs to be align with planned mappers in OIDC modules and we don't have them yet.

> I don't mind much waiting for https://github.com/quarkusio/quarkus/issues/39185 before finalizing this PR but completing https://github.com/quarkusio/quarkus/pull/39815 may take a while and it would be more difficult afterwards to move some of the builder API support for OidcCommonConfig.

I think it depends on your time to iterate over `@ConfigMapping` builders, it's better not to introduce brand new client reg. config and later migrate them, but as long as migration path is smooth.

So: this one is not getting into 3.13 anyway, we can replace config classes with builder before next release and have dedicated PR where we can only discuss how the builder should looks like. I'll leave decision to you.

**Comment by sberyozkin** [comment]
Thanks @michalvavrik

**Comment by sberyozkin** [comment]
@michalvavrik I'd probably go for keeping the current solution and like I said, we'd just get rid of `OidcClientRegistrationConfig` `ConfigGroup`, when the rest of the OIDC config hierarchy is ready, as we have open hands with the experimental status of this extension and a clear message that it can change. I can't really start with the `OidcClientRegistrationConfig` builder now because this PR moves some of the `OidcCommonConfig` to `OidcCommonClientConfig`...

I'll be going on PTO in a couple of days so we can finalize when I get back after 12th Aug, thanks

**Comment by michalvavrik** [comment]
> @michalvavrik I'd probably go for keeping the current solution and like I said, we'd just get rid of `OidcClientRegistrationConfig` `ConfigGroup`, when the rest of the OIDC config hierarchy is ready, as we have open hands with the experimental status of this extension and a clear message that it can change. I can't really start with the `OidcClientRegistrationConfig` builder now because this PR moves some of the `OidcCommonConfig` to `OidcCommonClientConfig`...

IMO makes sense; though you need to convince George as well :-)
 
> I'll be going on PTO in a couple of days so we can finalize when I get back after 12th Aug, thanks

yeah, have something that will stop me from contributing until then, let's get back on this after 12th, enjoy!

**Comment by sberyozkin** [comment]
Sure, thanks @michalvavrik. I might have an intermediate plan.

* 1st phase:  Remove `@ConfigGroup` from all the `OidcTenantConfig`, and `OidcClientConfig` hierarchy, make OidcCommonConfig abstract, and keep them for users who already instantiate them directly, while adding parallel ConfigMappings for `OidcCommonConfig, `OidcTenantConfig`
*  2nd phase: we work on the builder support for `OidcCommonConfig, `OidcTenantConfig`,  `OidcClientConfig` and deprecate a direct use of  them
* 3rd phase longer term: we remove `OidcCommonConfig, `OidcTenantConfig`,  `OidcClientConfig` classes

As far as this PR is concerned, I guess it may be simpler to merge it, once I can get an approval from Pedro as well, and then immediately afterwards follow up with the 1st phase, since this PR splits `OidcCommonConfig`  which will impact `quarkus-oidc`, `quarkus-oidc-client`, but also, the corresponding builder API will be impacted if phases 1,2 go first. Then move with the 2nd phase but replace `OidcClientRegistrationConfig` introduced with this PR immediately - because this extension will remain experimental for a while.

George, @gastaldi, how does it sound to you ?

**Comment by gastaldi** [comment]
@sberyozkin sounds good to me

**Comment by sberyozkin** [comment]
Thanks @gastaldi.
By the way, @michalvavrik, do you recall if a `@ConfigMapping` interface can extend another `@ConfigMapping` interface, like ConfigGroup classes can ? If not it would be a prerequisite for the phase 2 :-)

File: docs/src/main/asciidoc/security-openid-connect-client-registration.adoc
Comment (pedroigor) [inline docs/src/main/asciidoc/security-openid-connect-client-registration.adoc:74]: I would rather remove this paragraph because the mechanism described there is not covered by the specs but is specific to Keycloak.

File: docs/src/main/asciidoc/security-openid-connect-client-registration.adoc
Comment (pedroigor) [inline docs/src/main/asciidoc/security-openid-connect-client-registration.adoc:113]: What if we only have `OidcClientRegistration` and use a qualifier to inject a specific client registration? In case of injecting multiple instances, devs could use an injection point of  `Instance` ?

File: docs/src/main/asciidoc/security-openid-connect-client-registration.adoc
Comment (pedroigor) [inline docs/src/main/asciidoc/security-openid-connect-client-registration.adoc:135]: What if we make it possible to create a `OidcTenantConfig` from the metadata as a built-in capability in `quarkus-oidc`? The code here should repeat a lot when using custom tenants based on registered clients.

Perhaps, even the the `application_type` claim from the client metadata can be used to infer the `ApplicationType` for the quarkus application.

File: docs/src/main/asciidoc/security-openid-connect-client-registration.adoc
Comment (pedroigor) [inline docs/src/main/asciidoc/security-openid-connect-client-registration.adoc:208]: Do you have plans to somehow store the registration access token so that the client can update itself in the future? I don't know how it would work but if the Quarkus application is using some storage (e.g.: JPA) and is able to store the token.

**Review by pedroigor — COMMENTED** [review]
@sberyozkin I'm sharing my initial feedback about the functionality you are proposing. It looks great and a very useful tool for those using OIDC. Mainly when automating deployments.

**Comment by pedroigor** [comment]
From a user perspective, I think you are addressing well the main use cases I can think of:

* Automatically register clients so that they can start interacting with an OP
* Register clients on-demand

The first use case above is a bit tricky in regards to the registration access token (if not using an initial access token or a bearer token, as Keycloak supports) because you'll need to store this token somehow in order to be able to manage the clients after the registration. But I guess it will be up to the application to provide mechanisms to store this token.

**Comment by sberyozkin** [comment]
Thanks @pedroigor, let me comment

File: docs/src/main/asciidoc/security-openid-connect-client-registration.adoc
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-openid-connect-client-registration.adoc:113]: @pedroigor Definitely we can do it, I've been thinking about doing it in phase 2. similarly to what was done for OidcClient.
Note `OidcClientRegistrations` covers 2 cases:
* Get access to individual `OidcClientRegistration` configurations done in `application.properties` and indeed, here, we can evolve it further to allow named injections, I did not do it straight away to keep the changes as minimum as possible and let the API settle a bit
* Accept new `OidcClientRegistration` configurations dynamically, for example, like in a case mentioned by @calvernaz, where a 3rd party client submits Client metadata to Quarkus which then initiates actual registrations to new realms etc

File: docs/src/main/asciidoc/security-openid-connect-client-registration.adoc
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-openid-connect-client-registration.adoc:135]: +1, in fact the initial draft PR content did try to do it, but then I thought it was getting too complex for a single PR.

Note, right now, the new `oidc-client-registration` extension is totally independent of `quarkus-oidc`, users can build a system application which only facilitates the client registration, makes the new client registration details visible to `quarkus-oidc` aware services via some medium like secret engines, which is what we touched on with @calvernaz, etc.

But yeah, I also had another extension prototyped, `oidc-client-registration-config-source` where, at the startup, the client is registered and `quarkus-oidc` has all its properties like client id, client secret, auto-resolved from those metadata. It was in the early draft PR, but then I thought, let's do it afterwards, once the core client registration API settles, and then we can make more integral to `quarkus-oidc`

File: docs/src/main/asciidoc/security-openid-connect-client-registration.adoc
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-openid-connect-client-registration.adoc:208]: @pedroigor It is already supported :-).

`OidcClientRegistration` gives `RegisteredClient` representing the already registered client which keeps that cient specific registration token. One of the test `TenantConfigResolver` uses it to [re-read](https://github.com/quarkusio/quarkus/pull/41866/files#diff-6154f623d750cd5f9dde30f4e8c7cbce822efa06ddd9aaf6e54228eb96a7a204R55) the given `RegisteredClient`'s metadata, update did not work for Keycloak, well it does work, but there is a strange side-effect impacting ID tokens (and the good news, it was already fixed in Keycloak, https://github.com/keycloak/keycloak/issues/31062, so I'll try the update a bit later one 25.0.3 is out).
But yeah, full CRUD is supported

Comment (sberyozkin) [reply docs/src/main/asciidoc/security-openid-connect-client-registration.adoc:208]: I haven't looked yet into persisting these client specific registration tokens though, hmm, as indeed, after the restart, it will be gone. 
I think it will be up to users how to persist them, but we do need to let users recreate already registered `RegisteredClient`, I guess, to start with, I'll add a few builder methods to create a `RegisteredClient` instance from the existing `ClientMetadata` and the reg token...

**Comment by sberyozkin** [comment]
Hi @pedroigor 

>The first use case above is a bit tricky in regards to the registration access token (if not using an initial access token or a bearer token, as Keycloak supports) because you'll need to store this token somehow in order to be able to manage the clients after the registration. Otherwise, everytime you start the application a new client will be created and I'm unsure if that is something you usually expect. But I guess it will be up to the application to provide mechanisms to store this tok

Right, as far as the reg token which allows to create new client registrations is concerned, it will be managed by the application, same way it manages preconfigured client id and secret.

But you raised a good point about registration tokens for the already registered clients, I'll definitely need to do some initial work to support restoring already `registered clients` after a restart, instead of trying to register them again. I'll work on this and then ask yourself and George to have a look.

As far as the initial registration is concerned, without any initial reg tokens, it depends on how a given OIDC provider supports it, for Keycloak it is about enabling some policies in realms - which is what this PR's custom realm does.

In both cases, with or without the initial reg token, the question of how t handle registered clients after a restart remains, and some early solution must be provided before PR can be approved, so I'll look into it next, cheers

**Comment by sberyozkin** [comment]
@gastaldi @pedroigor 

I've fixed the tests and added new content to the docs, showing a simple example of how an already registered client can read its own metadata, ex, when it needs to refresh it and I'll try update with Keycloak 25.0.3 or later.

More importantly, I've added one more doc section suggesting a strategy for avoiding the duplicate client registrations at startup in cases when client metadata are preconfigured:
* Delay the registration until it is necessary with a `register-early=false` property
* On shutdown: persist in DB, cache, the client's registration URI and token
* On startup: check if the  registration URI and token for a given client are available, if yes, read the existing registration, otherwise initiate the delayed new client registration

Other good ideas from Pedro, 1) support named `OidcClientRegistration` injections 2) Have a closer integration with `quarkus-oidc`, can definitely be looked at, I'm just not sure we should do it now, given the overall complexity, I'd rather get users experiment with it for a while first. The 2nd item will also require a transition from `ConfigRoot` to `ConfigMapping`

Thanks

**Comment by sberyozkin** [comment]
Hey @pedroigor @gastaldi 

With Keycloak 25.0.4, I was able to update the test to show the individual registered client's update also works, see the code starting from [here](https://github.com/quarkusio/quarkus/pull/41866/files#diff-6154f623d750cd5f9dde30f4e8c7cbce822efa06ddd9aaf6e54228eb96a7a204R53).

So we check there first that the registered client's name is `Default Client`, and then we update the name to `Default Client Updated` which is expected in the test [here](https://github.com/quarkusio/quarkus/pull/41866/files#diff-702497716af1be91190d6486e4b8caf82db9bad851b7fcf81c5a2011389d6e01R33).

That should get enough space for users to start experimenting. The extension is clearly marked as experimental even in the docs, so we'll have an easy way to tune the API for a while. Once this one is merged, I'll open follow up enhancement requests: named OidcClientRegistration injection, and closer integration with `quarkus-oidc`, which was in the original PR draft but like I said I thought it looked a bit raw and I decided not to try to do too much immediately - it feels to me the dynamic OIDC client registration is really an async admin process, unrelated to the work of quarkus-oidc itself, though we can try to do something nice for some simple quarkus-oidc scenarios later.

File: extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/runtime/OidcClientRegistrationImpl.java
Comment (gastaldi) [inline extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/runtime/OidcClientRegistrationImpl.java:204]: `log.debug` here for debugging purposes?

**Review by gastaldi — APPROVED** [review]
LGTM

File: extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/runtime/OidcClientRegistrationImpl.java
Comment (sberyozkin) [reply extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/runtime/OidcClientRegistrationImpl.java:204]: @gastaldi Sure :-), let me check all those catch blocks

File: docs/src/main/asciidoc/security-openid-connect-client-registration.adoc
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-openid-connect-client-registration.adoc:205]: Just an update here, a doc section has been added suggesting a strategy for avoiding duplicate registrations, `RegisteredClient` has had 2 methods added for users be able to persist the client specific reg URL and token.

**Comment by sberyozkin** [comment]
Hi @gastaldi @pedroigor 
 
I'm getting ready to merge it, as Pedro may not be have time. I'll definitely thought follow up with enhancement requests based on Pedro's feedback.

In the very short term I'd like users to try it and I think I may need to tweak ClientMetadata a bit to make it easier to avoid having to deal with JSON and let users add array properties to the builder, etc. It is too complex now to keep adding new things to it at the moment IMHO :-)

**Review by pedroigor — APPROVED** [review]

**Comment by michalvavrik** [comment]
FYI I have experienced failures in my PR here https://github.com/quarkusio/quarkus/pull/42935 that are caused by this. When I reverted this PR locally, it's fixed. No need to reaction, I just wanted to link it here. I'll comment more in https://github.com/quarkusio/quarkus/pull/42935.

**Comment by sberyozkin** [comment]
Hi @michalvavrik That was fixed in response to #43106

**Comment by michalvavrik** [comment]
> Hi @michalvavrik That was fixed in response to #43106

it was fixed here https://github.com/quarkusio/quarkus/pull/42947, nevertheless, it works. thanks for the fix

**Pull request merged**
